File: javascript/031/main.js
/*
 * ProjectEuler 031
 * JavaScript / main.js
 * copyright (c) 2015 Susisu
 *
 * Node.js v5.0.0
 * `node main.js`
 */

"use strict";

function main() {
    var coins = [200, 100, 50, 20, 10, 5, 2, 1];
    console.log(calc(coins, 200));
}

function calc(coins, rest) {
    if (rest === 0) {
        return 1;
    }
    else if (coins.length === 0) {
        return 0;
    }
    else if (rest >= coins[0]) {
        return calc(coins, rest - coins[0]) + calc(coins.slice(1), rest);
    }
    else {
        return calc(coins.slice(1), rest);
    }
}

main();

File: ocaml/031/main.ml
(*
    ProjectEuler 031
    OCaml / main.ml
    copyright (c) 2016 Susisu

    OCaml 4.02.3
    `ocamlfind ocamlopt -o bin/main main.ml`
*)

let coins = [200; 100; 50; 20; 10; 5; 2; 1]

let rec patterns rest = function
      [] when rest = 0        -> 1
    | []                      -> 0
    | (x :: xs) when x > rest -> patterns rest xs
    | (x :: xs) as cs         -> patterns rest xs + patterns (rest - x) cs

let _ =
    print_int (patterns 200 coins);
    print_newline ()
